### ISSUE TITLE
Issues with Chinese charaters

### ISSUE THREAD

**Issue by c856450** [pekingduck/metadsf#5]
I have problems with Chinese characters, e.g.

`$ metadsf -t 20__夏日戀人.dsf`
`TALB=In Brasil (2016, SACD, �K�)`
`TRCK=1/11`
`TIT2=�K�`
`TDRC=1989`
`TPE1=�s�`

The tags are properly displayed with other tools like foorbar2000.

When I remove and re-tag the DSF file with metadsf, it can be dispalyed with metadsf -t:

`$ metadsf --remove-tags=TPE1,TIT2 20__夏日戀人.dsf`
`$ metadsf --set-tag=TPE1="梅艳芳" --set-tag=TIT2="夏日戀人" 20__夏日戀人.dsf`
`$ metadsf -t 20__夏日戀人.dsf`
`TALB=In Brasil (2016, SACD, �K�)`
`TRCK=1/11`
`TDRC=1989`
`TIT2=夏日戀人`
`TPE1=梅艳芳`

However, the tags of the DSF are displayed in foobar2000 mis-encoded characters, see below:

![grafik](https://user-images.githubusercontent.com/25209078/121758243-e8334c00-cb20-11eb-9746-3875f0cfa254.png)

My locale is set to de_UTF-8. I do not have any issues with Chinese characters using metaflac.

How do I fix this issue?

**Comment by dgheath** [comment]
I fixed a similar issue by changing MetaDSF::printTags() as
std::cout << (*it)->frameID() << "=" << (*it)->toString().to8Bit(true) << std::endl;
You might give that a try.

**Comment by c856450** [comment]
Many thanks. I will give it a try.

**Comment by c856450** [comment]
I am not a coder. Please let me know where shall I apply the change, the following line in `src/metadsf.cpp`:

`void MetaDSF::printTags(const char *prefix) const`

If yes, in what? Replacing the line with the following seems not to be right:

`std::cout << (*it)->frameID() << "=" << (*it)->toString().to8Bit(true) << std::endl;`

**Comment by dgheath** [comment]
Yes, that is the correct place. The to8Bit(true) tells it to convert the string to UTF-8. I'm not familiar enough with the specifics to know if that is the correct fix for you. UTF-8 should be able to handle any language. Without this change, I think it was outputting latin1 which cannot handle many languages. Would be nice if the author could take a look at this. Good luck.

**Comment by zvezdochiot** [comment]
Hi @dgheath .

Your recipe will only help to display the contents of the tag. But not for input. This is half the solution. And the other half is unknown to me.

**Comment by c856450** [comment]
> Your recipe will only help to display the contents of the tag. But not for input. This is half the solution. And the other half is unknown to me.

@dgheath Yes, I can confirm. The display issue has been fixed with the change, i.e. `metadsf -t` works now. But the input issue using `metadsf --set-tag` still exists.

**Comment by zvezdochiot** [comment]
Hi all (@c856450 , @dgheath ).
```sh
$ ./metadsf -sTIT2=Тест test.dsf
```
simple view (not Ok)
```sh
$ ./metadsf -t test.dsf 
TALB=Depeche Mode
TCON=Other
TDRC=2006-05-02
TRCK=1
TIT2=Ð¢ÐµÑÑ
```
`iconv` convert (Ok)
```sh
$ ./metadsf -t test.dsf | iconv -t LATIN1
TALB=Depeche Mode
TCON=Other
TDRC=2006-05-02
TRCK=1
TIT2=Тест
```
Not work `-e{LATIN1,UTF8,UTF16}`!

**Comment by zvezdochiot** [comment]
Hi @c856450 .

For UTF8 small dirty patch.
https://github.com/pekingduck/metadsf/blob/eb01fb5b80f8eb7c269c4402e871e1021f5d521e/src/metadsf.cpp#L156-L163
to
```cpp
int MetaDSF::setTag(const TagLib::String &key, const TagLib::String &val, 
		    bool replace) 
{
  std::string valt = val.to8Bit(false);
  TagLib::String valu = TagLib::String(valt, _i->_encoding);
  TagLib::StringList vals;
  
  vals.append(valu);
  return setTag(key, vals, replace);
}
```
Result:
```sh
$ ./metadsf -sTIT2=Тест test.dsf
$ ./metadsf -t test.dsf 
TALB=Depeche Mode
TCON=Other
TDRC=2006-05-02
TRCK=1
TIT2=Тест
```

**Comment by c856450** [comment]
@zvezdochiot Wow! Wonderfull, this works now! Thank you so much, both!!!


The original `metadsf.cpp` seems to have considered the encoding. However, as pointed by @zvezdochiot `-e{LATIN1,UTF8,UTF16}` does not work.
